### ISSUE TITLE
Plugin set development improvements

### ISSUE THREAD

**Issue by Enet4** [bioinformatics-ua/dicoogle#115]
These are a few ideas that I had some months ago. They will not be too complicated to implement and will certainly make plugins look cleaner, safer and easier to develop. I am leaving the details below so that it can be considered for integration in the future.

The plugin set is meant to be an aggregation of the plugins that we which to inject to the system, and right now it feels occluded by a multitude of methods that need to be implemented before things work. We can give our developers the confidence that they are in charge of their plugins while keeping things simple.

##### Proposal Part 1: Make the development of a plugin set based on annotations rather than implementing an interface.

The key concept here is that instead of having plugin getters (one method per interface type), the plugins will be declared as non-static attributes of the plugin set class. New annotations can be used to retrieve these instances, as well as to identify the plugin set. Only the default constructor will be required (how this is achieved is further explained in the next parts).

`@DicooglePluginSet(String value)` should extend jspf's `@PluginImplementation` and is also used to define the name of the plugin set (`getName` is discarded). Then we can have one annotation for each kind of interface (`@QueryPlugin`, `@IndexerPlugin`, `@StoragePlugin`, `@ServletPlugin`, `@RestletPlugin`, ...). No collections are created and all plugin types are optional.

```Java
@DicooglePluginSet("RSI")
public class RSIPluginSet {
  @QueryPlugin
  final RSIQueryInterface queryInt;
  @IndexerPlugin
  final RSIIndexerInterface indexer;

  // other plugins here, no getters

  public RSIPluginSet() {
    // initialize plugins here, as usual
  }
}
```

##### Proposal part 2: Settings are automatically propagated to each piece.

Rather than having the plugin set send the configuration holder manually to each plugin, we should let the platform do it instead. Ultimately, only the instances need these settings. With that in mind, their propagation should be handled by the core, and not the set.

##### Proposal part 3: The platform interface is also propagated to each piece.

We can stick to the concept of `PlatformCommunicatorInterface`, this time leaving the propagation to the core.

The `setPlatformProxy` method that needs to be implemented behaves like a handler of the event _"the platform is ready to serve this plugin"_. Therefore, we could optionaly rename it to something like `onPlatformReady(DicooglePlatformInterface)`, which would likely make it more understandable.

##### Proposal part 4:  No `shutdown` at plugin set level.

I have looked into the code of multiple plugins and found out that none of them is using it. There is a good reason to this too: the plugins themselves should handle shutdown events, not their composer. If any resources have to be cleaned, they must be handled by the plugins involved, and the platform will do that instead.

----------------------------

These are the reasons why I think we should support these changes:

1. _It's simpler and more intuitive._ Picturing the diverse range of plugin developers, who might not be programming experts, I count this as a plus. It will also help expert developers at understanding what should and should not be done.

2. _It's safer._ We have faced issues before with an implementation of a plugin set that was not thread-safe. With plugin getters gone, that no longer poses an issue. We can also make sure that we would always have the same plugins in a set, regardless of when they are retrieved. We know this is only performed on start-up, but some plugin developers may think that a reassignment is possible when it really isn't.

3. _It's extensible._ Whether it is appreciated or not, the requirements for next-generation PACS archives might change along with Dicoogle's. With this approach, compatibility with previous plugins is not broken in the event that new plugin interfaces are introduced.

**Comment by tmgodinho** [comment]
Sounds like a very cool idea. I think it will help a lot to develop the plugins, especially new developers.
Can't wait to use :)

**Comment by bastiao** [comment]
Nice improvement to PluginBase: https://github.com/bioinformatics-ua/dicoogle/blob/dev/sdk/src/main/java/pt/ua/dicoogle/sdk/PluginBase.java 

We will analyse and discuss it in a meeting after release 2.0. 

Good job.

**Comment by Enet4** [comment]
Yes, good thing you've reminded me of PluginBase. I've updated the proposal to make it clear that both `PluginSet` and `PluginBase` become obsolete.

**Comment by Enet4** [comment]
I have revised the proposal again in order to fix another flaw. We cannot load plugins if they do not implement a `Plugin` interface.

**Comment by Enet4** [comment]
I have updated the proposal. After some rethinking, I figured that plugin-scoped resources still make sense.

**Comment by Enet4** [comment]
While these ideas are interesting, it is too late for something like this to enter 3.0.0. Maybe Dicoogle 4? :)
